### PR TITLE
Add template get function

### LINKED ISSUE #21: The Service type should be define separately
Currently, a Service is a principal,  but in order to generalize over getters and constructors, 
it would be handy if it is defined  in its own class, a simple solution would be:

`class Service: Principal {
     ...
}`

The service is used interchangeably as a Principal. 
these characteristics appeal to an inheritance approach from the Principal class.

:link: [zboto Link](https://app.clickup.com/t/861n00h3y)

### PR DESCRIPTION
**Pull request by neithanmo**
This define a generic get function to retrieve values from the inner IdlValue class, this implementation supports:
- all primitives types integer, bool, float and nats.
- string
- IDLValue::Number
- Principal
- Func
- Vec<IdlValue>
- Service

not supported yet: 
- Record, see #19 
- Variant, see #11
- Opt

fixes #21 #22 #23

:link: [zboto Link](https://app.clickup.com/t/861n01h7n)

### CODE REVIEW

**Comment by becominginsane** [comment]
I think the record shouldn't be blocked by #19.
What do we define as ergonomic to get the record? I think it's ok to have a dedicated method instead to disambiguate entirely what the getter should be doing.
That, or a `CandidRecord` wrapper class that we can use to create and manage product types would be useful. We could `.get<CandidRecord>()` and inspect the fields with it, or conversely use a builder to create a record correctly and use it with `IdlValue(CandidRecord)`

File: lib-agent-cpp/src/idl_value.cpp [hunk 1]
```diff
@@ -149,99 +161,53 @@ IdlValue IdlValue::FromFunc(std::vector<uint8_t> vector,
...
 
-std::string IdlValue::getText() {
+template <>
+std::optional<Number> IdlValue::get() {
   if (ptr == nullptr) {
-    std::cerr << "IdlValue instance uninitialized" << std::endl;
-    return "";
+    return std::nullopt;
   }
   CText *ctext = text_from_idl_value(ptr);
```
Comment (becominginsane) [inline lib-agent-cpp/src/idl_value.cpp]: Nullable?

File: lib-agent-cpp/src/idl_value.cpp [hunk 2]
```diff
@@ -250,54 +216,78 @@ std::string IdlValue::getText() {
...
 
-std::optional<zondax::Principal> IdlValue::getPrincipal() {
+template <>
+std::optional<zondax::Func> IdlValue::get() {
   if (ptr == nullptr) {
-    std::cerr << "IdlValue instance uninitialized" << std::endl;
     return std::nullopt;
   }
-  CPrincipal *p = principal_from_idl_value(ptr);
+  struct CFunc *cFunc = func_from_idl_value(ptr);
```
Comment (becominginsane) [inline lib-agent-cpp/src/idl_value.cpp:230]: more nullable

**Comment by neithanmo** [comment]
latest changes:
- fix Number api using ctext  #23 
- Add Service type definition and specialize a getter for it. fixes #21

**Review by becominginsane — APPROVED** [review]

**Review by chcmedeiros — APPROVED** [review]

**Pull request merged**